### PR TITLE
[PTCD-864] Add repeated name fields back to provider search index to fix rankings

### PR DESCRIPTION
**Pull request by timabell**
The original search index had all four of the name fields, which meant that the name had a higher weight in the result scores.

* This is a follow-up to the creation of the new index PR #1936 
* The front-end changes to use the new index are in PR #1956

[PTCD-864](https://skillsfundingagency.atlassian.net/browse/PTCD-864)

sql index ranking success Selection_20210304-01-7eda17eda17eda1
(showing changes with #1956 applied locally and this PR deployed to dev environment)

### CODE REVIEW

**Comment by timabell** [comment]
[release 1216](https://sfa-gov-uk.visualstudio.com/Digital%20First%20Careers/_releaseProgress?releaseId=18732&_a=release-pipeline-progress) for testing on dev

**Comment by timabell** [comment]
For this to correctly update the existing index we need to manually delete the old one because the DevOps scripts don't currently handle altering indexes.

delete previous deployment of index Selection_20210224-01-1af651af651af65

**Comment by timabell** [comment]
[release 1217](https://sfa-gov-uk.visualstudio.com/Digital%20First%20Careers/_releaseProgress?releaseId=18737&_a=release-pipeline-progress) take 2 of dev test deployment, skipped non-search steps in deploy last time but actually need the updated cosmos/sql sync deployed to get the new data.

**Comment by timabell** [comment]
[release 1218](https://sfa-gov-uk.visualstudio.com/Digital%20First%20Careers/_releaseProgress?releaseId=18744&_a=release-pipeline-progress) for testing updated search on dev azure search instance

**Comment by timabell** [comment]
Much closer now, though still not identical ordering

**Review by soharadesign — APPROVED** [review]

**Comment by timabell** [comment]
[release 1227](https://sfa-gov-uk.visualstudio.com/Digital%20First%20Careers/_releaseProgress?releaseId=18784&_a=release-pipeline-progress) for testing, now with even more unused fields

**Comment by timabell** [comment]
[release 1228](https://sfa-gov-uk.visualstudio.com/Digital%20First%20Careers/_releaseProgress?releaseId=18785&_a=release-pipeline-progress) for test deploy to dev (just dacpac and searches)

**Comment by timabell** [comment]
[release 1230](https://sfa-gov-uk.visualstudio.com/Digital%20First%20Careers/_releaseProgress?releaseId=18793&_a=release-pipeline-progress) full release, try again

File: src/Dfc.CourseDirectory.Database/Pttcd/Views/vwProviderSearchIndex.sql [hunk 1]
```diff
@@ -3,11 +3,19 @@ AS
...
             p.ProviderName,
+            ISNULL(pc.AddressPostTown, pc.AddressItems) Town,
+            pc.AddressPostcode Postcode,
             p.ProviderStatus,
             p.UkrlpProviderStatusDescription,
-            pc.AddressPostcode Postcode,
-            ISNULL(pc.AddressPostTown, pc.AddressItems) Town,
-            p.Version -- Timestamp (RowVersion) type, used here for cosmos HighWaterMark
+            p.Version, -- Timestamp (RowVersion) type, used here for cosmos HighWaterMark
+
```
Comment (jameslawesynergy) [inline src/Dfc.CourseDirectory.Database/Pttcd/Views/vwProviderSearchIndex.sql:11]: Why lb?

File: src/Dfc.CourseDirectory.Database/Pttcd/Views/vwProviderSearchIndex.sql [hunk 2]
```diff
@@ -3,11 +3,19 @@ AS
...
+            -- additional fields to skew the search index score approximately as per the old index
+            coalesce(p.CourseDirectoryName,p.ProviderName) CourseDirectoryName, -- as per previous cosmos-based search index definition
```
Comment (jameslawesynergy) [inline src/Dfc.CourseDirectory.Database/Pttcd/Views/vwProviderSearchIndex.sql:13]: Use CAPS for function names, space after commas for readability

File: src/Dfc.CourseDirectory.Database/Pttcd/Views/vwProviderSearchIndex.sql [hunk 3]
```diff
@@ -3,11 +3,19 @@ AS
...
+            p.TradingName,
+            p.Alias,
+            p.UpdatedOn,
+            cast(null as datetime) as DateOnboarded,
```
Comment (jameslawesynergy) [inline src/Dfc.CourseDirectory.Database/Pttcd/Views/vwProviderSearchIndex.sql:17]: Use CAPS for function names and `NULL` for readability

File: src/Dfc.CourseDirectory.Database/Pttcd/Views/vwProviderSearchIndex.sql [hunk 4]
```diff
@@ -3,11 +3,19 @@ AS
...
+            cast(null as varchar(255)) as Region
```
Comment (jameslawesynergy) [inline src/Dfc.CourseDirectory.Database/Pttcd/Views/vwProviderSearchIndex.sql:18]: Inconsistent convention for specifying column names, use `AS` keyword for all, or none. (Also CAPS preferable)

**Comment by timabell** [comment]
rebased

File: src/Dfc.CourseDirectory.Database/Pttcd/Views/vwProviderSearchIndex.sql (hunk 1 above)
Comment (timabell) [reply src/Dfc.CourseDirectory.Database/Pttcd/Views/vwProviderSearchIndex.sql:11]: Sorry I don't understand the question, could you elaborate please

Comment (jameslawesynergy) [reply src/Dfc.CourseDirectory.Database/Pttcd/Views/vwProviderSearchIndex.sql:11]: Line break

File: src/Dfc.CourseDirectory.Database/Pttcd/Views/vwProviderSearchIndex.sql (hunk 3 above)
Comment (timabell) [reply src/Dfc.CourseDirectory.Database/Pttcd/Views/vwProviderSearchIndex.sql:17]: I will reformat the file for consistency with everything else we have as suggested.

As an aside, personally I'd drop the all-caps use of sql across the board having stopped to think about why it's like that in the first place and how that affects readability. It's not required by the parser (as of sql-99). It's just a hangover from a language that was created in 1974 and everyone is just so used to reading and writing it that way that it's the norm. After finding it a bit jarring the first few times I did it all in lowercase I very quickly started to find it easier to read.

Related:

* https://stackoverflow.com/questions/9426138/why-are-the-queries-in-sql-mostly-written-in-capital-letters
* https://www.mity.com.au/blog/writing-readable-content-and-why-all-caps-is-so-hard-to-read

File: src/Dfc.CourseDirectory.Database/Pttcd/Views/vwProviderSearchIndex.sql (hunk 2 above)
Comment (timabell) [reply src/Dfc.CourseDirectory.Database/Pttcd/Views/vwProviderSearchIndex.sql]: done, thanks

File: src/Dfc.CourseDirectory.Database/Pttcd/Views/vwProviderSearchIndex.sql (hunk 4 above)
Comment (timabell) [reply src/Dfc.CourseDirectory.Database/Pttcd/Views/vwProviderSearchIndex.sql]: done, thanks

File: src/Dfc.CourseDirectory.Database/Pttcd/Views/vwProviderSearchIndex.sql (hunk 1 above)
Comment (timabell) [reply src/Dfc.CourseDirectory.Database/Pttcd/Views/vwProviderSearchIndex.sql:11]: Oh I see. I want to group the fields after the linebreak together as a "just to fix the rankings" section. I couldn't find a way of making that clear that I liked.
Without the line break it looks like the comment only applies to the immediately following field and not all the rest. I'll have another go at making it clear without being ugly, if you have any suggestions that'd be grand.

Comment (timabell) [reply src/Dfc.CourseDirectory.Database/Pttcd/Views/vwProviderSearchIndex.sql:11]: Or this the option of just leaving it in the git blame and not commenting it at all, but I think I'd like to highlight that these fields are only for ranking.
Another option is to add a comment to every ranking-only field, which might actually be tidier than this.

Comment (timabell) [reply src/Dfc.CourseDirectory.Database/Pttcd/Views/vwProviderSearchIndex.sql]: Okay, I've pushed something I'm happer with. Thanks for the prompt. Take a look and see what you think.

**Review by jameslawesynergy — APPROVED** [review]
NB requires manual release steps

**Comment by timabell** [comment]
> NB requires manual release steps

These have been documented in the jira ticket (this is how we agreed to make sure these happen). Please have a look there and see if you think I've missed anything @jameslawesynergy

**Comment by timabell** [comment]
[make a new search index for provider that points at mssql](https://trello.com/c/2CGlxL5t/43-make-a-new-search-index-for-provider-that-points-at-mssql)

**Pull request merged**